### PR TITLE
merge bitcoin#15596...#16542: backports

### PR DESCRIPTION
**Pull request by kittywhiskers**

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -878,8 +878,8 @@ static UniValue sendmany(const JSONRPCRequest& request)
                             {"address", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "The dash address is the key, the numeric amount (can be string) in " + CURRENCY_UNIT + " is the value"},
                         },
                     },
-                    {"minconf", RPCArg::Type::NUM, /* default */ "1", "Only use the balance confirmed at least this many times."},
-                    {"addlocked", RPCArg::Type::BOOL, /* default */ "false", "Whether to include transactions locked via InstantSend."},
+                    {"minconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED_NAMED_ARG, "Ignored dummy value"},
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:880]: 15596: missing `if (request.fHelp ...` move

File: test/functional/test_framework/test_framework.py [hunk 2]
```diff
@@ -645,23 +645,14 @@ def skip_if_no_cli(self):
...
-
-        return config["components"].getboolean("ENABLE_WALLET")
+        return self.config["components"].getboolean("ENABLE_WALLET")
 
     def is_zmq_compiled(self):
         """Checks whether the zmq module was compiled."""
-        config = configparser.ConfigParser()
-        config.read_file(open(self.options.configfile))
-
         return config["components"].getboolean("ENABLE_ZMQ")
```
Comment (UdjinM6) [inline test/functional/test_framework/test_framework.py:656]: 15895: `self.config`

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2163,90 +2164,90 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
...
+            tip_hash = locked_chain->getBlockHash(*tip_height);
+        }
+        block_height = locked_chain->getBlockHeight(block_hash);
+        progress_begin = chain().guessVerificationProgress(block_hash);
+        progress_end = chain().guessVerificationProgress(stop_block.IsNull() ? tip_hash : stop_block);
+    }
+    double progress_current = progress_begin;
+    while (block_height && !fAbortRescan && !chain().shutdownRequested()) {
+        if (*block_height % 100 == 0 && progress_end - progress_begin > 0.0) {
+            ShowProgress(strprintf("%s " + _("Rescanning..."), GetDisplayName()), std::max(1, std::min(99, (int)((progress_current - progress_begin) / (progress_end - progress_begin) * 100))));
```
Comment (UdjinM6) [inline src/wallet/wallet.cpp:2272]: 15901:
```suggestion
        m_scanning_progress = (progress_current - progress_begin) / (progress_end - progress_begin);
        if (*block_height % 100 == 0 && progress_end - progress_begin > 0.0) {
            ShowProgress(strprintf("%s " + _("Rescanning..."), GetDisplayName()), std::max(1, std::min(99, (int)(m_scanning_progress * 100))));
```

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -119,6 +119,7 @@ UniValue importprivkey(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_WALLET_ERROR, "Cannot import private keys to a wallet with private keys disabled");
     }
 
+    WalletBatch batch(pwallet->GetDBHandle());
```
Comment (UdjinM6) [inline src/wallet/rpcdump.cpp:122]: 15741: lots of unrelated(?) changes

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1453,6 +1453,15 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
     return true;
 }
 
+void RelayTransaction(const uint256& txid, const CConnman& connman)
```
Comment (UdjinM6) [inline src/net_processing.cpp:1456]: 16452: missing some previous PR

File: test/functional/rpc_psbt.py [hunk 6]
```diff
@@ -69,6 +69,15 @@ def run_test(self):
         assert_equal(walletprocesspsbt_out['complete'], True)
         self.nodes[1].sendrawtransaction(self.nodes[1].finalizepsbt(walletprocesspsbt_out['psbt'])['hex'])
 
+        # feeRate of 0.1 BTC / KB produces a total fee slightly below -maxtxfee (~0.05280000):
+        res = self.nodes[1].walletcreatefundedpsbt([{"txid":txid,"vout":p2wpkh_pos},{"txid":txid,"vout":p2sh_p2wpkh_pos},{"txid":txid,"vout":p2pkh_pos}], {self.nodes[1].getnewaddress():29.99}, 0, {"feeRate": 0.1})
```
Comment (UdjinM6) [inline test/functional/rpc_psbt.py:73]: 16257: s/p2wpkh_pos/p2pkh_pos/ here and below

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/wallet/wallet.h [hunk 7]
```diff
@@ -1002,7 +1002,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
     };
     ScanResult ScanForWalletTransactions(const uint256& first_block, const uint256& last_block, const WalletRescanReserver& reserver, bool fUpdate);
     void TransactionRemovedFromMempool(const CTransactionRef &ptx, MemPoolRemovalReason reason) override;
-    void ReacceptWalletTransactions();
+    void ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/wallet/wallet.h:970]: That means this entire PR is blocked by https://github.com/dashpay/dash/pull/4568, move commit or let PR stall?

File: src/wallet/rpcdump.cpp (hunk 4 above)
Comment (kittywhiskers) [reply src/wallet/rpcdump.cpp:122]: Batching requires the presence of `WalletBatch batch(pwallet->GetDBHandle());` which hasn't been defined and following the spirit of https://github.com/bitcoin/bitcoin/pull/15741/commits/ccb26cf347a6fcc861768bbfbbd652e683ffd5bb, the changes were permeated wherever necessary

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (kittywhiskers) [reply src/wallet/rpcwallet.cpp:880]: `RPCHelpMan` AFAIK is not implemented so it was skipped

File: test/functional/rpc_psbt.py (hunk 6 above)
Comment (kittywhiskers) [reply test/functional/rpc_psbt.py]: Done!

File: test/functional/rpc_psbt.py [hunk 8]
```diff
@@ -69,6 +69,15 @@ def run_test(self):
         assert_equal(walletprocesspsbt_out['complete'], True)
         self.nodes[1].sendrawtransaction(self.nodes[1].finalizepsbt(walletprocesspsbt_out['psbt'])['hex'])
 
+        # feeRate of 0.1 BTC / KB produces a total fee slightly below -maxtxfee (~0.05280000):
```
Comment (UdjinM6) [inline test/functional/rpc_psbt.py:72]: 16257: 
- dashify
- fees are higher in our case - `0.06650000` https://gitlab.com/dashpay/dash/-/jobs/1861193316#L2158
- should also tweak `assert_greater_than`s below accordingly

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/rpc_psbt.py (hunk 8 above)
Comment (kittywhiskers) [reply test/functional/rpc_psbt.py]: Resolved in b2ca69100c40d1cb9e75aa3fb954ab1e1499a858, thanks a lot!

File: src/net_processing.cpp (hunk 5 above)
Comment (kittywhiskers) [reply src/net_processing.cpp:1452]: Earliest records point to [`#9260`](https://github.com/bitcoin/bitcoin/pull/9260/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R843-R850) so backporting is out of scope.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -2784,46 +2784,6 @@ CAmount CWallet::GetNormalizedAnonymizedBalance() const
     return nTotal;
 }
 
-// Calculate total balance in a different way from GetBalance. The biggest
```
Comment (PastaPastaPasta) [inline src/wallet/wallet.cpp:2788]: 15596: You should also be doing the pcoin -> wtx here where possible no? At least one example is https://github.com/dashpay/dash/blob/06cc4bbe5f1364800a66b31c61a4794f3fe90128/src/wallet/wallet.cpp#L4365

Comment (kittywhiskers) [reply src/wallet/wallet.cpp:2788]: I've attempted to use https://github.com/bitcoin/bitcoin/pull/15596/commits/faa3a246e8809b4954a4a6d202fe8c7f7f776b6e and it breaks builds for some reason

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see 946ca478b58a133b33dad8a107314e733965ed5b and 3a70b5aa01b9c446d4a617d32d74ff0a905d6f87

**Review by UdjinM6 — APPROVED** [review]
LGTM, lightly tested ACK

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**